src/analyses/memLeak.ml

```diff
@@ -19,6 +19,22 @@ struct
   let context _ _ = ()
 
   (* HELPER FUNCTIONS *)
+  let get_global_vars () =
+    (* Filtering by GVar seems to account for declarations, as well as definitions of global vars *)
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
+
+  let get_reachable_mem_from_globals (global_vars:varinfo list) ctx =
+    global_vars
+    |> List.map (fun v -> Lval (Var v, NoOffset))
+    |> List.filter_map (fun exp ->
```

Comment:
This can be improved a little: If we have global structs for example, we should follow all the pointer fields of the struct.

Reply:
Thanks for the idea! I quickly tried out a few regr. tests with global struct variables (both pointer and non-pointer ones, as well as with nested structs; cf. `76/{10, 11, 12}`)

It seems that without changing anything in the implementation, Goblint can already detect the `valid-memtrack` violations in these test cases.

Nonetheless, if you think it makes sense to add anything further to the implementation, I'm more than happy to do that.

Reply:
> It seems that without changing anything in the implementation, Goblint can already detect the `valid-memtrack` violations in these test cases.

Great! I think the other interesting case would be the cases where `valid-memtrack` holds, because something is still reachable via a global struct member. Have you also tried that direction?

Reply:
If it means reachable, then it should use `Queries.ReachableFrom`.

Reply:
No, it is **must**-reachable, while the query is **may**-reachable.

Reply:
The current `MayPointTo` query is also a may-query. The must-ness comes from the singleton check of the may result.

Maybe the non-transitive aspect of `MayPointTo` is fine here and must-reachability via pointers would just make us less precise? In that case the function should also be named accordingly to avoid the inconsistency.

Reply:
Must-reachability by pointers would make us more precise: One would consider all the things that the globals (or it's struct members) must point to (i.e. singleton points-to-set), and then can consider these as roots as well to include such things that must-be-reachable from these.
Doing a MayReachable query and then only considering the set if it is a singleton set would be less precise than considering the MayPointTo set if it is singleton.

Reply:
> Great! I think the other interesting case would be the cases where valid-memtrack holds, because something is still (edit: must-)reachable via a global struct member. Have you also tried that direction?

I added some changes with respect to this suggestion. Please check out my comment here: https://github.com/goblint/analyzer/pull/1241#issuecomment-1813364110